src/Lean/Widget/InteractiveDiagnostic.lean

```diff
@@ -28,30 +28,41 @@ inductive MsgEmbed where
   | expr : CodeWithInfos → MsgEmbed
   /-- An interactive goal display. -/
   | goal : InteractiveGoal → MsgEmbed
-  /-- Some messages (in particular, traces) are too costly to print eagerly. Instead, we allow
-  the user to expand sub-traces interactively. -/
+  /-- A widget instance.
+
+  `alt` is a fallback rendering of the widget
+  that can be shown in standard, non-interactive LSP diagnostics,
+  as well as when user widgets are not supported by the client. -/
+  | widget (wi : Widget.WidgetInstance) (alt : TaggedText MsgEmbed)
+  /-- Some messages (in particular, traces) are too costly to print eagerly.
+  Instead, we allow the user to expand sub-traces interactively. -/
   | trace (indent : Nat) (cls : Name) (msg : TaggedText MsgEmbed) (collapsed : Bool)
       (children : StrictOrLazy (Array (TaggedText MsgEmbed)) (WithRpcRef LazyTraceChildren))
   deriving Inhabited, RpcEncodable
 
-/-- The `message` field is the text of a message possibly containing interactive *embeds* of type
-`MsgEmbed`. We maintain the invariant that embeds are stored in `.tag`s with empty `.text` subtrees,
-i.e. `.tag embed (.text "")`, because a `MsgEmbed` display involve more than just text. -/
+/-- The `message` field is the text of a message
+possibly containing interactive *embeds* of type `MsgEmbed`.
+We maintain the invariant that embeds are stored in `.tag`s with empty `.text` subtrees,
+i.e., `.tag embed (.text "")`.
+
+Client-side display algorithms render tags in a custom way,
+ignoring the nested text. -/
 abbrev InteractiveDiagnostic := Lsp.DiagnosticWith (TaggedText MsgEmbed)
 
 deriving instance RpcEncodable for Lsp.DiagnosticWith
 
 namespace InteractiveDiagnostic
 open MsgEmbed
 
-def toDiagnostic (diag : InteractiveDiagnostic) : Lsp.Diagnostic :=
+partial def toDiagnostic (diag : InteractiveDiagnostic) : Lsp.Diagnostic :=
```

Comment:
Afaict yes, since `prettyTt` calls itself inside `TaggedText.rewrite` now. One could inline `rewrite`, but that also is `partial` without `List/Array.pmap` which is not in core.

Reply:
Ah, I didn't realize that it's not in core. Unfortunate!